File: Java/lib/src/test/java/leetcode/LongestCommonPrefix/LongestCommonPrefixTest.java
package leetcode.LongestCommonPrefix;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.stream.Stream;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

class LongestCommonPrefixTest {
  protected Solution solution = new Solution();

  static Stream<Arguments> provideTestCases() {
    return Stream.of(
        Arguments.of("common prefix", new String[] {"flower", "flow", "flight"}, "fl"),
        Arguments.of("no common prefix", new String[] {"dog", "racecar", "car"}, ""),
        Arguments.of("empty string array", new String[] {}, ""),
        Arguments.of("single string", new String[] {"single"}, "single"),
        Arguments.of("all strings are the same", new String[] {"test", "test", "test"}, "test"),
        Arguments.of("mixed case strings", new String[] {"a", "A"}, ""),
        Arguments.of("one empty string", new String[] {"", "b"}, ""));
  }

  @ParameterizedTest(name = "{0}")
  @MethodSource("provideTestCases")
  void testLongestCommonPrefix(String name, String[] strs, String want) {
    String actual = solution.longestCommonPrefix(strs);
    assertEquals(want, actual);
  }
}

File: Go/longest_common_prefix/longest_common_prefix_test.go
package longest_common_prefix

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestLongestCommonPrefix(t *testing.T) {
	cases := []struct {
		name string
		strs []string
		want string
	}{
		{
			name: "common prefix exists",
			strs: []string{"flower", "flow", "flight"},
			want: "fl",
		},
		{
			name: "empty string slice",
			strs: []string{},
			want: "",
		},
		{
			name: "no common prefix",
			strs: []string{"dog", "racecar", "car"},
			want: "",
		},
		{
			name: "single string",
			strs: []string{"alone"},
			want: "alone",
		},
		{
			name: "no common prefix at all",
			strs: []string{"abc", "def", "ghi"},
			want: "",
		},
		{
			name: "same strings",
			strs: []string{"same", "same", "same"},
			want: "same",
		},
		{
			name: "prefix is the shortest string",
			strs: []string{"ab", "abc", "abcd"},
			want: "ab",
		},
	}

	for _, tt := range cases {
		t.Run(tt.name, func(t *testing.T) {
			actual := longestCommonPrefix(tt.strs)
			assert.Equal(t, actual, tt.want)
		})
	}
}
